From 85ea20a819fb5f73433835266893c12bb08bbf8e Mon Sep 17 00:00:00 2001
From: William Douglas <william.douglas@intel.com>
Date: Wed, 7 Feb 2018 21:33:30 +0000
Subject: [PATCH] Allow detection of Clear Linux OS in python3

In python3 the platform.linux_distribution() function does not display
content for Clear Linux. In addition the linux_distribution function
is due to be deprecated in python 3.7 (the next minor release, see
https://bugs.python.org/issue1322). In anticipation of this, switch to
sourcing the os-release file in the waagent.service and reading the
distribution information from the environment instead.
---
 azurelinuxagent/common/version.py | 17 ++++++++++++++++-
 init/clearlinux/waagent.service   |  1 +
 2 files changed, 17 insertions(+), 1 deletion(-)

diff --git a/azurelinuxagent/common/version.py b/azurelinuxagent/common/version.py
index 890a9af..76a3a0c 100644
--- a/azurelinuxagent/common/version.py
+++ b/azurelinuxagent/common/version.py
@@ -58,6 +58,7 @@ def get_f5_platform():
 
 
 def get_distro():
+    osinfo = None
     if 'FreeBSD' in platform.system():
         release = re.sub('\-.*\Z', '', ustr(platform.release()))
         osinfo = ['freebsd', release, '', 'freebsd']
@@ -67,9 +68,23 @@ def get_distro():
                                                   supported_dists=supported))
         full_name = platform.linux_distribution()[0].strip()
         osinfo.append(full_name)
-    else:
+    elif 'dist' in dir(platform):
         osinfo = platform.dist()
 
+    # In the case that platform.dist doesn't provide this information
+    # (see https://bugs.python.org/issue1322)
+    # take the details from the environment for Clear Linux (as the
+    # waagent.service will add them).
+    #
+    # Eventually consider using a library to handle other distros like:
+    # https://pypi.python.org/pypi/distro which is built to do this as
+    # the stdlib folks prefer this functionality live in pypi.
+    if not osinfo or osinfo[0] == '' and os.environ.get('NAME'):
+        osinfo = [os.environ['NAME'],
+                  os.environ['VERSION_ID'],
+                  os.environ['ID'],
+                  os.environ['PRETTY_NAME']]
+
     # The platform.py lib has issue with detecting oracle linux distribution.
     # Merge the following patch provided by oracle as a temporary fix.
     if os.path.exists("/etc/oracle-release"):
diff --git a/init/clearlinux/waagent.service b/init/clearlinux/waagent.service
index 9afee45..b2d0e08 100644
--- a/init/clearlinux/waagent.service
+++ b/init/clearlinux/waagent.service
@@ -8,6 +8,7 @@ ConditionPathExists=/usr/share/defaults/waagent/waagent.conf
 
 [Service]
 Type=simple
+EnvironmentFile=/usr/lib/os-release
 ExecStart=/usr/bin/python -u /usr/bin/waagent -daemon
 Restart=always
 RestartSec=5
-- 
2.16.1

